### PR TITLE
[wip] Use utest for our unit tests

### PR DESCRIPTION
**Pull request by Simn**
Several problems still, keeping this open to keep track of the progress.

## neko

```
src/unit/TestMain.hx:51: START
assertations: 5023
successes: 5022
errors: 1
failures: 0
warnings: 0
execution time: 0.317
results: SOME TESTS FAILURES (success: false)
unit.TestReflect
  testTypes: ERROR ...........E
    $hiter
Called from haxe.ds.StringMap::keys line 50
Called from haxe.ds.StringMap::toString line 73
Called from <header> line 1
Called from a C function
Called from Std::string line 34
Called from unit.Test::eq line 22
Called from unit.TestReflect::testTypes line 108
Called from unit.TestReflect::__initializeUtest__ line 96
Command exited with 1 in 0s: neko [bin/unit.n]
```

I have no idea why it's calling `Std::string` from `Test::eq`, or why that fails in the process.

## js

We have to "port" this piece of code to utest:

```
	#if js
	static function nodejsMain() {
		main();
		(untyped process).exit(Test.success ? 0 : 1);
	}
	#end
```

I couldn't immediately find how to get the success state of the tests. @RealyUniqueName

## as3

```
/home/travis/build/Simn/haxe/tests/unit/bin/as3/utest/ITestHandler.as(103): col: 31 Error: Incompatible override.
		protected override function bindHandler() : void {
```

This is some Haxe problem related to the `@:public` metadata being added by `can_access`. However, after fixing that there are some more problems. It would be good if we could find a way to dodge this in utest. @RealyUniqueName 

## lua

```
Command exited with 0 in 4s: haxe [compile-lua.hxml,-D,travis]
Command: lua [bin/unit.lua]
src/unit/TestMain.hx:49: Generated at: 2019-02-28 19:55:51
src/unit/TestMain.hx:51: START
utest/ui/text/PrintReport.hx:52: 
assertations: 5045
successes: 5044
errors: 1
failures: 0
warnings: 0
execution time: 1.067
results: SOME TESTS FAILURES (success: false)
unit.TestSerialize
  test: ERROR .....................................................................................................................................................................................................................................................E
    bin/unit.lua:70629: stack overflow
Command exited with 1 in 1s: lua [bin/unit.lua]
test lua failed
```

Probably something silly, though I don't know why changing the test frameworks brings this up. @jdonaldson 

## cppia

see #7898

### CODE REVIEW

File: tests/unit/spamfest [hunk 1]
```diff
@@ -0,0 +1,1586 @@
+test-adapter is recording results...
+
+src/unit/TestMain.hx:50: Generated at: 2019-02-28 18:20:30
```
Comment (Gama11) [inline tests/unit/spamfest]: Did you mean to push this? :)

Comment (Simn) [reply tests/unit/spamfest]: Oops, thanks for spotting that!

**Comment by Simn** [comment]
Dodged neko and fixed the original as3 problem. As3 still fails to compile this utest code:

```haxe
var init:TestData.InitializeUtest = (cast testCase:TestData.Initializer).__initializeUtest__();
```

The problem is that `TestData.Initializer` is a typedef, so the as3 generator doesn't know how to cast it. Let's please just dodge this in utest (e.g. by assigning to a var `#if as3`). @RealyUniqueName

**Comment by Simn** [comment]
There's a high chance that the lua problem comes from a to-string operation. Unfortunately, utest is very eager to stringify values in its assertion functions:

```haxe
  public static function equals(expected : Dynamic, value : Dynamic, ?msg : String , ?pos : PosInfos) {
    if(msg == null) msg = "expected " + q(expected) + " but it is " + q(value);
    isTrue(expected == value, msg, pos);
  }
```

IMO this is not good design, but on the other hand we shouldn't end up with stack overflows regardless.

**Comment by Simn** [comment]
Yep:

```haxe
class Main {
	var ref:Main;

	public function new() {}

	static public function main() {
		var m = new Main();
		m.ref = m;
		trace(Std.string(m));
	}
}
```

**Pull request merged**